src/layer_env.rs

```diff
@@ -0,0 +1,882 @@
+                }
+            }
+        };
+
+        target_delta.insert(modification_behavior, name, value);
+    }
+
+    /// Constructs a `LayerEnv` based on the given layer directory.
+    ///
+    /// Follows the rules described in the Cloud Native Buildpacks specification and adds implicit
+    /// entries for special directories (such as `bin`) should they exist.
+    ///
+    /// **NOTE**: Buildpack authors should **never directly use this** in their buildpack code and
+    /// rely on libcnb to pass `LayerEnv` values to minimize side effects in buildpack code.
+    ///
+    /// # Example:
+    ///```
+    /// use libcnb::layer_env::{LayerEnv, TargetLifecycle};
+    /// use tempfile::tempdir;
+    /// use libcnb::Env;
+    /// use std::fs;
+    ///
+    /// // Create a bogus layer directory
+    /// let temp_dir = tempdir().unwrap();
+    /// let layer_dir = temp_dir.path();
+    /// fs::create_dir_all(layer_dir.join("bin")).unwrap();
+    ///
+    /// let layer_env_dir = layer_dir.join("env");
+    /// fs::create_dir_all(&layer_env_dir).unwrap();
+    /// fs::write(layer_env_dir.join("ZERO_WING.default"), "ALL_YOUR_BASE_ARE_BELONG_TO_US").unwrap();
+    ///
+    /// let layer_env = LayerEnv::read_from_layer_dir(&layer_dir).unwrap();
+    ///
+    /// let env = Env::new();
+    /// let modified_env = layer_env.apply(TargetLifecycle::Launch, &env);
+    ///
+    /// assert_eq!(modified_env.get("PATH").unwrap(), layer_dir.join("bin"));
+    /// assert_eq!(modified_env.get("ZERO_WING").unwrap(), "ALL_YOUR_BASE_ARE_BELONG_TO_US");
+    /// ```
+    pub fn read_from_layer_dir(layer_dir: impl AsRef<Path>) -> Result<LayerEnv, std::io::Error> {
+        let mut result_layer_env = LayerEnv::new();
+
+        let bin_path = layer_dir.as_ref().join("bin");
+        let lib_path = layer_dir.as_ref().join("lib");
+        let include_path = layer_dir.as_ref().join("include");
+        let pkgconfig_path = layer_dir.as_ref().join("pkgconfig");
+
+        let layer_path_specs = vec![
+            ("PATH", TargetLifecycle::Build, &bin_path),
+            ("LIBRARY_PATH", TargetLifecycle::Build, &lib_path),
+            ("LD_LIBRARY_PATH", TargetLifecycle::Build, &lib_path),
+            ("CPATH", TargetLifecycle::Build, &include_path),
+            ("PKG_CONFIG_PATH", TargetLifecycle::Build, &pkgconfig_path),
+            ("PATH", TargetLifecycle::Launch, &bin_path),
+            ("LD_LIBRARY_PATH", TargetLifecycle::Launch, &lib_path),
+        ];
+
+        for (name, target_lifecycle, path) in layer_path_specs {
+            if path.is_dir() {
+                let target_delta = match target_lifecycle {
+                    TargetLifecycle::Build => &mut result_layer_env.layer_paths_build,
+                    TargetLifecycle::Launch => &mut result_layer_env.layer_paths_launch,
+                    _ => panic!("Unexpected TargetLifecycle in read_from_layer_dir implementation. This is a libcnb implementation error!"),
```

Comment:
This should be unreachable if I'm reading right because we control all the inputs. If not then I was about to suggest we format the target_lifecycle and put it in the error message.

Reply:
Correct, that's why the panic says "This is a libcnb implementation error!". To be honest I'm not super happy about this code. But adding another enum just for this case felt excessive.